### PR TITLE
Feat: add traffic mesh proxy flag

### PR DESCRIPTION
**Pull request by erezrokah**
Fixes https://github.com/netlify/traffic-mesh/issues/806

This is a big PR since I did some refactoring to the dev command in the process since it was very hard to follow.
The commits are small though.

I'll add comments to clarify the changes.

This PR consists of 3 main changes:
1. Extract common logic to download binary clients (this was used in the tunnel client and now shared with traffic mesh agent).
2. Extract the different dev command responsibilities (starting functions server, framework server, Netlify dev proxy, tunnel client, etc) into separate functions/files.
3. Traffic Mesh flag implementation which was quite small.

> When enabling traffic mesh some test are still failing - looking into those.

### CODE REVIEW

File: src/lib/exec-fetcher.js [hunk 1]
```diff
@@ -0,0 +1,90 @@
+const fs = require('fs-extra')
```
Comment (erezrokah) [inline src/lib/exec-fetcher.js:1]: Most of the logic here was extracted from https://github.com/netlify/cli/blob/93468a3fc43db5d637b30b9cbb21042da8626962/src/utils/live-tunnel.js

And made generic so we can re-use it.

File: src/lib/exec-fetcher.js [hunk 2]
```diff
@@ -0,0 +1,90 @@
...
+  }
+
+  try {
+    const outdated = await isVersionOutdated({
+      packageName,
+      currentVersion: match[1],
+    })
+    return outdated
+  } catch (e) {
+    if (exists) {
```
Comment (erezrokah) [inline src/lib/exec-fetcher.js:68]: Added a small fix to use existing version if checking for updates failed.
That can happen when reaching GitHub API limit:
https://github.com/netlify/gh-release-fetch/pull/3

File: src/lib/exec-fetcher.test.js [hunk 3]
```diff
@@ -0,0 +1,70 @@
...
+test(`should postix exec with .exe on windows`, t => {
+  const execName = 'some-binary-file'
+  if (process.platform === 'win32') {
+    t.is(getExecName({ execName }), `${execName}.exe`)
+  } else {
+    t.is(getExecName({ execName }), execName)
+  }
+})
+
+const packages = [
```
Comment (erezrokah) [inline src/lib/exec-fetcher.test.js:22]: Feels a bit strange to have this here and not in a dedicated test file (one for the tunnel client and one for the traffic mesh).
However I prefer it this way since we're actually testing the `exec-fetcher`

File: src/lib/settings.js [hunk 4]
```diff
@@ -0,0 +1,15 @@
+const path = require('path')
```
Comment (erezrokah) [inline src/lib/settings.js:1]: I've started a new concept of `libs` instead of `utils` that we currently have.
It seems that most of our current `utils` are not used for sharing code, but just to extract functionality from the different commands.
However some commands/utils need to share the same code - so I started putting those parts under lib.

File: src/utils/live-tunnel.js [hunk 5]
```diff
@@ -58,70 +55,38 @@ async function connectTunnel(session, netlifyApiToken, localPort, log) {
...
-  return stat && stat.isFile() && isExe(stat.mode, stat.gid, stat.uid)
+    extension: process.platform === 'win32' ? 'zip' : 'tar.gz',
+  })
 }
 
-function isExe(mode, gid, uid) {
-  if (isWindows()) {
-    return true
-  }
+const startLiveTunnel = async ({ siteId, netlifyApiToken, localPort, log }) => {
```
Comment (erezrokah) [inline src/utils/live-tunnel.js:80]: `createTunnel` and `connectTunnel` were called separately in the dev command.
I grouped them into  `startLiveTunnel`

File: src/utils/proxy.js [hunk 6]
```diff
@@ -0,0 +1,346 @@
+const path = require('path')
```
Comment (erezrokah) [inline src/utils/proxy.js:1]: This was mostly copied as is from the `dev` command

File: src/commands/dev/index.js [hunk 7]
```diff
@@ -384,8 +29,11 @@ async function startDevServer(settings, log) {
       },
     })
 
-    server.start(function() {
-      log(`\n${NETLIFYDEVLOG} Server listening to`, settings.frameworkPort)
+    await new Promise(resolve => {
```
Comment (erezrokah) [inline src/commands/dev/index.js:32]: Added some code to wait until the server is ready

File: src/utils/proxy.js [hunk 8]
```diff
@@ -0,0 +1,346 @@
...
+      return proxy.web(req, res, options)
+    },
+    ws: (req, socket, head) => proxy.ws(req, socket, head),
+  }
+
+  return handlers
+}
+
+async function startProxy(settings = {}, addonUrls, configPath, projectDir) {
+  const functionsServer = settings.functionsPort ? `http://localhost:${settings.functionsPort}` : null
```
Comment (erezrokah) [inline src/utils/proxy.js:286]: The original `startProxy` had some code to wait for the framework server port and functions server port:
https://github.com/netlify/cli/blob/93468a3fc43db5d637b30b9cbb21042da8626962/src/commands/dev/index.js#L170
https://github.com/netlify/cli/blob/93468a3fc43db5d637b30b9cbb21042da8626962/src/commands/dev/index.js#L178

I removed it in favour of:
1. Using `waitPort` in the function that starts the framework server: https://github.com/netlify/cli/pull/1196/files#diff-8a2f65e3aede3ea9696e546766f50f03R82
2. When starting the functions server - wait on a promise until the server is ready: https://github.com/netlify/cli/pull/1196/files#diff-cf2bd94d7688f0711bbd931e6c5e0397R373

File: src/utils/proxy.js [hunk 9]
```diff
@@ -0,0 +1,346 @@
...
+
+      proxy.web(req, res, options)
+    })
+  })
+
+  server.on('upgrade', function(req, socket, head) {
+    proxy.ws(req, socket, head)
+  })
+
+  return new Promise(resolve => {
```
Comment (erezrokah) [inline src/utils/proxy.js:339]: Also added this code to wait for the dev server to be ready

File: src/utils/proxy.js [hunk 10]
```diff
@@ -0,0 +1,346 @@
...
+    server.listen({ port: settings.port }, () => {
+      resolve(`http://localhost:${settings.port}`)
```
Comment (erezrokah) [inline src/utils/proxy.js:341]: The original `startProxy` returned an object with `url` and `port` as properties:
https://github.com/netlify/cli/blob/93468a3fc43db5d637b30b9cbb21042da8626962/src/commands/dev/index.js#L235
but only `url` was used.
Simplified the function to return just the `url` as a string

File: src/utils/serve-functions.js [hunk 11]
```diff
@@ -318,4 +310,77 @@ async function serveFunctions(dir, siteInfo = {}) {
   return app
 }
 
-module.exports = { serveFunctions }
+const getBuildFunction = ({ functionBuilder, log }) => {
```
Comment (erezrokah) [inline src/utils/serve-functions.js:313]: Moved this chunk of code from the `dev` command

File: src/utils/serve-functions.js [hunk 12]
```diff
@@ -318,4 +310,77 @@ async function serveFunctions(dir, siteInfo = {}) {
...
+        functionWatcher.on('add', debouncedBuild)
+        functionWatcher.on('change', debouncedBuild)
+        functionWatcher.on('unlink', debouncedBuild)
+      })
+    }
+
+    const functionsServer = await serveFunctions(settings.functions, siteInfo)
+
+    await new Promise(resolve => {
+      functionsServer.listen(settings.functionsPort, err => {
```
Comment (erezrokah) [inline src/utils/serve-functions.js:374]: Added this to wait until the server is ready

File: src/utils/traffic-mesh.js [hunk 13]
```diff
@@ -0,0 +1,70 @@
+const path = require('path')
```
Comment (erezrokah) [inline src/utils/traffic-mesh.js:1]: This is very similar to the tunnel client code

File: src/utils/traffic-mesh.js [hunk 14]
```diff
@@ -0,0 +1,70 @@
...
+    process.on(signal, () => {
+      const sig = signal == 'exit' ? 'SIGTERM' : signal
+      subprocess.kill(sig, {
+        forceKillAfterTimeout: 2000,
+      })
+    })
+  )
+
+  try {
+    const open = await waitPort({ port, output: 'silent', timeout: 30 * 1000 })
```
Comment (erezrokah) [inline src/utils/traffic-mesh.js:60]: Wait for the agent to be ready (the traffic mesh agent writes to a log file and not stdout) so waiting on the port works well.

File: tests/command.dev.test.js [hunk 15]
```diff
@@ -6,118 +6,104 @@ const FormData = require('form-data')
 const { withSiteBuilder } = require('./utils/siteBuilder')
 const { startExternalServer } = require('./utils/externalServer')
 
-test('should return index file when / is accessed', async t => {
-  await withSiteBuilder('site-with-index-file', async builder => {
-    builder.withContentFile({
-      path: 'index.html',
-      content: '<h1>⊂◉‿◉つ</h1>',
-    })
+const testMatrix = [
```
Comment (erezrokah) [inline tests/command.dev.test.js:9]: The changes in this file is for support a test matrix with different args (and creating unique test names)

**Review by ehmicky — APPROVED** [review]
:rocket:

**Comment by erezrokah** [comment]
@ehmicky wow that was fast. Going to disable the new failing test (they are failing to due GitHub API limit reached on CI).
